src/libos/src/events/waiter_queue_observer.rs

```diff
@@ -0,0 +1,35 @@
+use std::any::Any;
+use std::marker::PhantomData;
+
+use super::{Event, Observer, WaiterQueue};
+use crate::prelude::*;
+
+/// A Observer associated with a WaiterQueue.
+///
+/// Once the observer receives any interesting events, it will dequeue and
+/// wake up all `Waiters` in the associated `WaiterQueue`.
+pub struct WaiterQueueObserver<E: Event> {
```

Comment:
I _deliberately_ keep `Waiter`/`Waker`/`WaiterQueue` and `Event`/`Observer`/`Notifier` separated. It makes the code conceptually simple (each file is less than or around 100 lines of code), clear, and decoupled from each other (`Waiter` has no idea of `Event`, and `WaiterQueue` does know what is a `Notifier`). These primitives can be used individually (e.g., futex only needs to use `Waiter`-family, not `Event`-family primitives) or can be combined in various ways (e.g., `WaiterQueueObserver`), powerful enough to express what you refer to as "`WaitQueue` as an observee of the observer".

Reply:
Then I prefer `Arc<WaiterQueue>`. I think an observe should not own its observee. Using a reference seems extensible.

Reply:
An interesting suggestion. But, no, you don't want to implement `Observer<E>` for `WaiterQueue`. The problem is to implement the `Observer` trait, one has to provide an implementation of `on_event` method. But, what should the desirable behavior for `WaiterQueue` to do when receiving events? It depends on the specific situation. Sometimes, it is to wake up all waiters (this is exactly what `WaiterQueueObserver` does); sometimes, it could wake just one waiters; sometimes, maybe a specific number of waiters. There seems to be no universal answer to that. Maybe as more use cases are implemented upon the event subsystem, we can eventually find the answer. But before that, I think it is better off keeping the flexibility.